snuba/subscriptions/data.py

```diff
@@ -181,16 +192,23 @@ def from_dict(
             resolution_sec=int(data["resolution"]),
             query=data["query"],
             entity_subscription=entity_subscription,
+            metadata=metadata,
         )
 
     def to_dict(self) -> Mapping[str, Any]:
-        return {
+        subscription_data_dict = {
             "project_id": self.project_id,
             "time_window": self.time_window_sec,
             "resolution": self.resolution_sec,
             "query": self.query,
-            **self.entity_subscription.to_dict(),
         }
+        if self.entity_subscription.processors:
+            for processor in self.entity_subscription.processors:
+                subscription_data_dict.update(processor.to_dict(self.metadata))
+        if self.entity_subscription.validators:
+            for validator in self.entity_subscription.validators:
+                subscription_data_dict.update(validator.to_dict(self.metadata))
```

Comment:
It's not a validators job to be return data to write to redis

Reply:
Why not? What if we want to validate based on metadata in the subscription?

Reply:
I think a validator needs to do only that - validate. That is provide a validation function that can return true if valid and false if not. So it's fine if it wants to validate subscription metadata but it is not ok that it is the one deciding what metadata to actually write there. A validator should never be mutate data in our systems. If it's doing more than validating, maybe it should not be called something else.

This is the reason why I suggest you do not split processors and validators as I think they might be inherently linked.

Reply:
@enochtangg If we removed the `to_dict` function from the validator and only had the processor write metadata, would that break any existing subscriptions?

Reply:
That makes sense. We can remove the `to_dict` abstract method from the validator base. Right now it only returns an empty dictionary.

Reply:
OK let's do that then. That keeps the logic in the processor alone.